src/components/product/dto/create-product.dto.ts

```diff
@@ -0,0 +1,58 @@
+import { Type } from 'class-transformer';
+import {
+  ValidateNested,
+} from 'class-validator';
+import { Field, InputType, ObjectType } from 'type-graphql';
+import { Product } from './product.dto';
+import { BibleBook } from './bible-book';
+import { ProductType } from './product-type';
+import { ProductMedium } from './product-medium';
+import { ProductApproach } from './product-approach';
+import { ProductMethodology } from './product-methodology';
+import { ProductPurpose } from './product-purpose';
+
+@InputType()
+export abstract class CreateProduct {
+  @Field(type => ProductType)
+  readonly type: ProductType;
+
+  @Field(type => [BibleBook])
+  readonly books: BibleBook[];
+
+  @Field(type => [ProductMedium])
+  readonly mediums: ProductMedium[];
+
+  @Field(type => [ProductPurpose])
+  readonly purposes: ProductPurpose[];
+
+  @Field(type => ProductApproach)
+  readonly approach: ProductApproach;
+
+  @Field(type => ProductMethodology)
+  readonly methodology: ProductMethodology;
+}
+
+// Used for GQL testing, where enum values must be key-based as opposed to value-based
+// e.g. in gql, type: 'bible_stories' should be type: 'BibleStories'
+export abstract class CreateProductRaw {
```

Comment:
This one threw me for a loop, wasn't happy about this approach. Couple options off the top of my head:

1. Change all enums to contain the same key/value. e.g. `BibleStories: 'BibleStories'` - (not sure the extent to which we're using enums throughout the backend codebase and the client-side, but it's probably too late to introduce a new standard like this)

2. Perhaps a utility could be created that takes a Class (e.g. `CreateProduct`) as an argument that will auto-convert all enum values from value->key?

Reply:
Yeah this also surprised me too. Typically enums are a mapping of a "private" key to a "public" value. In this case it's the opposite. 
The author does mention it [here](https://typegraphql.ml/docs/enums.html) under _Interoperability_.

For the most part, I'm much more concerned about the API definition than the DB schema. API crosses repo boundaries and can create "breaking changes", where the DB schema is "private" to the repo and can be changed without affecting the API (potentially).
We are still really early in development though so breaking changes anywhere are fine. In the future, they will be fine too with just a bit more thought and graceful rollout.

I think your first option is the best one to go with. Make the enum keys and values the same. This will create less work in tests and prevent surprises when expecting the key instead of value or vise-versa. It's also recommended by the library author lol. I also don't think we are concerned about the size of these strings in the db nor over the wire, which I think would be the only reason to do a mapping.

Reply:
Continuing in https://github.com/SeedCompany/cord-api-v3/issues/142